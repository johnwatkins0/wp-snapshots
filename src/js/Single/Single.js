/* eslint react/no-danger: 0 */
import React from 'react';
import PropTypes from 'prop-types';

import { Facebook, Twitter } from '../svg';

const Single = ({
  title: { rendered: title },
  content: { rendered: content },
  excerpt: { rendered: excerpt },
  image: {
    media_details: {
      sizes: { large: image },
    },
  },
}) => (
  <article className="SnapshotsSingle">
    <header className="SnapshotsSingle__header">
      <div className="SnapshotsSingle__title-container">
        <h1 className="SnapshotsSingle__title" dangerouslySetInnerHTML={{ __html: title }} />
      </div>
      <img
        src={image.source_url}
        width={image.width}
        height={image.height}
        alt={title}
        className="SnapshotsSingle__image"
      />
    </header>
    <div dangerouslySetInnerHTML={{ __html: content }} />
    <footer className="SnapshotsSingle__share-band">
      <h2>Share</h2>
      <a
        target="_blank"
        href={`https://www.facebook.com/sharer/sharer.php?u=${window.location.href}`}
      >
        <Facebook />
      </a>
      <a
        target="_blank"
        href={`https://twitter.com/intent/tweet?url=${window.location.href}&text=${excerpt}`}
      >
        <Twitter />
      </a>
    </footer>
  </article>
);

Single.propTypes = {
  title: PropTypes.objectOf(PropTypes.any),
  content: PropTypes.objectOf(PropTypes.any),
  excerpt: PropTypes.objectOf(PropTypes.any),
  image: PropTypes.objectOf(PropTypes.any).isRequired,
};

Single.defaultProps = {
  title: {
    rendered: '',
  },
  content: {
    rendered: '',
  },
  excerpt: {
    rendered: '',
  },
};

export default Single;
