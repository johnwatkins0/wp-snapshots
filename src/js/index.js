/* eslint react/no-danger: 0 */
import React, { Component } from 'react';
import { render } from 'react-dom';
import PropTypes from 'prop-types';
import { BrowserRouter, Route, Switch, Link } from 'react-router-dom';
import DocumentTitle from 'react-document-title';
import get from 'lodash.get';
import { parse } from 'qs';

import Archive from './Archive';
import Single from './Single';
import MiniNav from './MiniNav';

import { appDidMount } from './store/actions';
import effects from './store/effects';
import reducer from './store/reducer';
import { getPostBySlug, getPosts } from './store/selectors';

import { ROOT_SELECTOR } from './constants';
import { withScrollToTop } from './utils/withScrollToTop';
import { withReplaceEditLink } from './utils/withReplaceEditLink';

const ScrollToTopSingle = withScrollToTop(withReplaceEditLink(Single, 'post'));
const ScrollToTopArchive = withScrollToTop(withReplaceEditLink(Archive, 'category'));

const INITIAL_STATE = {
  posts: [],
};

export class App extends Component {
  static propTypes = {
    termId: PropTypes.string.isRequired,
    termName: PropTypes.string.isRequired,
    termSlug: PropTypes.string.isRequired,
    termDescription: PropTypes.string.isRequired,
    basenamePrefix: PropTypes.string,
    titleAppend: PropTypes.string,
  };

  static defaultProps = {
    basenamePrefix: '',
    titleAppend: '',
  };

  constructor(props) {
    super(props);

    this.state = INITIAL_STATE;
    this.dispatch = this.dispatch.bind(this);
    this.triggerEffect = this.triggerEffect.bind(this);
    this.getState = this.getState.bind(this);
  }

  componentDidMount() {
    this.dispatch(appDidMount(this.props.termId, this.props.previewPost || null));
  }

  getState = () => ({ ...this.state });

  dispatch(action) {
    this.setState(state => reducer(state, action), () => this.triggerEffect(action));
  }

  triggerEffect(action) {
    if (action.type in effects) {
      effects[action.type](action, {
        dispatch: this.dispatch,
        getState: this.getState,
      });
    }
  }

  render = () => (
    <BrowserRouter basename={this.props.basenamePrefix}>
      <main className="SnapshotsMain">
        <Switch>
          <Route
            path="/snapshots/:name/"
            render={() => (
              <DocumentTitle title={this.props.termName + this.props.titleAppend}>
                <ScrollToTopArchive
                  posts={getPosts(this.state)}
                  termName={this.props.termName}
                  termDescription={this.props.termDescription}
                  basename={this.props.basenamePrefix}
                  id={Number(this.props.termId)}
                />
              </DocumentTitle>
            )}
          />

          <Route
            path="/snapshot/:slug/"
            component={({
              match: {
                params: { slug },
              },
            }) => {
              const post = getPostBySlug(this.state, slug);
              return (
                <DocumentTitle
                  key={get(post, 'id', '')}
                  title={get(post, ['title', 'rendered'], document.title) + this.props.titleAppend}
                >
                  <div>
                    <div className="SnapshotsMain__title-container">
                      <h1>
                        <Link
                          href={`/snapshots/${this.props.termSlug}/`}
                          to={`/snapshots/${this.props.termSlug}/`}
                          dangerouslySetInnerHTML={{ __html: this.props.termName }}
                        />
                      </h1>
                    </div>
                    {post &&
                      post.image && (
                        <ScrollToTopSingle {...post} basename={this.props.basenamePrefix} />
                      )}
                    <MiniNav
                      posts={getPosts(this.state)}
                      termName={this.props.termName}
                      termSlug={this.props.termSlug}
                    />
                  </div>
                </DocumentTitle>
              );
            }}
          />

          <Route
            path="/"
            component={({ location }) => {
              const search = location.search || '';
              const params = parse(search.replace('?', ''));
              if (!('preview' in params) || !('p' in params)) {
                return null;
              }

              if (!this.props.previewPost) {
                return null;
              }

              const post = this.state.posts.find(({ id }) => id === Number(params.p));
              return (
                <DocumentTitle
                  key={get(post, 'id', '')}
                  title={get(post, ['title', 'rendered'], document.title) + this.props.titleAppend}
                >
                  <div>
                    <div className="SnapshotsMain__title-container">
                      <h1>
                        <Link
                          href={`/snapshots/${this.props.termSlug}/`}
                          to={`/snapshots/${this.props.termSlug}/`}
                          dangerouslySetInnerHTML={{ __html: this.props.termName }}
                        />
                      </h1>
                    </div>
                    {post &&
                      post.image && (
                        <ScrollToTopSingle {...post} basename={this.props.basenamePrefix} />
                      )}
                    <MiniNav
                      posts={getPosts(this.state)}
                      termName={this.props.termName}
                      termSlug={this.props.termSlug}
                    />
                  </div>
                </DocumentTitle>
              );
            }}
          />

        </Switch>
      </main>
    </BrowserRouter>
  );
}

export const startApp = () => {
  [...document.querySelectorAll(ROOT_SELECTOR)].forEach((root) => {
    render(
      <App
        termName={root.getAttribute('data-term-name')}
        titleAppend={root.getAttribute('data-title-append')}
        termId={root.getAttribute('data-term-id')}
        termSlug={root.getAttribute('data-term-slug')}
        termDescription={root.getAttribute('data-term-description')}
        basenamePrefix={root.getAttribute('data-basename-prefix').replace(/\/$/, '') || ''}
        previewPost={JSON.parse(root.getAttribute('data-preview-post'))}
      />,
      root,
    );
  });
};

export const start = async () => {
  if (!('fetch' in window)) {
    await import('whatwg-fetch');
  }

  startApp();
};

window.addEventListener('load', start);
