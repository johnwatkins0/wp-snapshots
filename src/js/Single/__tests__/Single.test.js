import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter, Route } from 'react-router-dom';

import { testPostsWithMedia } from '../../../../demo/src/data';

import Single from '..';

const post = testPostsWithMedia[0];

const Component = () => (
  <BrowserRouter>
    <Route>
      <Single {...post} />
    </Route>
  </BrowserRouter>
);

describe('Single', () => {
  it('renders', () => {
    const tree = renderer.create(<Component />).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
