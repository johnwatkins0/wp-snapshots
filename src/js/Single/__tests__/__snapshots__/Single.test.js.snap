// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Single renders 1`] = `
<article
  className="SnapshotsSingle"
>
  <header
    className="SnapshotsSingle__header"
  >
    <div
      className="SnapshotsSingle__title-container"
    >
      <h1
        className="SnapshotsSingle__title"
        dangerouslySetInnerHTML={
          Object {
            "__html": "Donec vel volutpat",
          }
        }
      />
    </div>
    <img
      alt="Donec vel volutpat"
      className="SnapshotsSingle__image"
      height={683}
      src="/demo/src/images/1024x683-7.jpg"
      width={1024}
    />
  </header>
  <div
    dangerouslySetInnerHTML={
      Object {
        "__html": "<section><p>Lorem ipsum dolor sit amet, consectetur adipiscing elit. Nunc quis libero at sapien tempus semper. Mauris quis ante velit. Donec vel volutpat sapien, eu efficitur mi. Etiam lacinia finibus erat, a porttitor massa auctor ut. Duis mauris nunc, sagittis non lobortis sed, volutpat quis dui. In accumsan blandit vestibulum. Aliquam erat volutpat.</p>
</section>",
      }
    }
  />
  <footer
    className="SnapshotsSingle__share-band"
  >
    <h2>
      Share
    </h2>
    <a
      href="https://www.facebook.com/sharer/sharer.php?u=about:blank"
      target="_blank"
    >
      <svg
        height="24px"
        viewBox="0 0 1792 1792"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Facebook
        </title>
        <path
          d="M1376 128q119 0 203.5 84.5t84.5 203.5v960q0 119-84.5 203.5t-203.5 84.5h-188v-595h199l30-232h-229v-148q0-56 23.5-84t91.5-28l122-1v-207q-63-9-178-9-136 0-217.5 80t-81.5 226v171h-200v232h200v595h-532q-119 0-203.5-84.5t-84.5-203.5v-960q0-119 84.5-203.5t203.5-84.5h960z"
          fill="currentColor"
        />
      </svg>
    </a>
    <a
      href="https://twitter.com/intent/tweet?url=about:blank&text=<p>Nunc quis libero</p>
"
      target="_blank"
    >
      <svg
        height="24px"
        viewBox="0 0 1792 1792"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
      >
        <title>
          Twitter
        </title>
        <path
          d="M1684 408q-67 98-162 167 1 14 1 42 0 130-38 259.5t-115.5 248.5-184.5 210.5-258 146-323 54.5q-271 0-496-145 35 4 78 4 225 0 401-138-105-2-188-64.5t-114-159.5q33 5 61 5 43 0 85-11-112-23-185.5-111.5t-73.5-205.5v-4q68 38 146 41-66-44-105-115t-39-154q0-88 44-163 121 149 294.5 238.5t371.5 99.5q-8-38-8-74 0-134 94.5-228.5t228.5-94.5q140 0 236 102 109-21 205-78-37 115-142 178 93-10 186-50z"
          fill="currentColor"
        />
      </svg>
    </a>
  </footer>
</article>
`;
